[CORE-13235] cloud_topics/reconciler: Use L1 io and object primitives

wdberkeley opened the pull request:
This alters the reconciler so it builds and uploads proper L1 objects, using the l1::io and l1::object_builder abstractions.

A follow-up will change the commit and LRO offset portions of the reconciler to use the metastore and the ctp_stm.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

wdberkeley committed:
cloud_topics/reconciler: Use L1 io and object primitives

This alters the reconciler so it builds and uploads proper L1 objects,
using the l1::io and l1::object_builder abstractions.

A follow-up will change the commit and LRO offset portions of the
reconciler to use the metastore and the ctp_stm.

dotnwat reviewed (commented):
took a quick look and no red flags, but it seems like there are several logical changes combined into one commit which could be factored out into separate commits.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,93 +157,145 @@ ss::future<> reconciler::reconciliation_loop() {
+          partition->partition->ntp(),
+          partition->lro);
+
         auto reader = co_await make_reader(partition, size_budget);
```
FYI this function has UAF and I will have a commit to fix it up tomorrow.

rockwotj commented on src/v/cloud_topics/reconciler/range_batch_consumer.h:
```diff
@@ -23,37 +23,26 @@
-/*
- * Consumer that builds a range from a record batch reader.
- */
+// TODO: Maybe this should be renamed?
```
Maybe something simple like `reconciliation_consumer`?

rockwotj replied:
A brief comment about this class would be helpful

rockwotj commented on src/v/cloud_topics/reconciler/range_batch_consumer.h:
```diff
@@ -23,37 +23,26 @@
+    l1::object_builder* _builder;
+    model::ntp _ntp;
+    partition_metadata _metadata;
     std::optional<kafka::offset> _base_offset;
```
I think it will be cleaner to instead represent the state as:

```
std::optional<partition_metadata> _metadata;
```

rockwotj commented on src/v/cloud_topics/reconciler/tests/range_batch_consumer_test.cc:
```diff
@@ -43,11 +58,27 @@ TEST(RangeBatchConsumer, MultipleBatchesRecords) {
-            ASSERT_EQ(range->info.base_offset(), base_offset);
-            ASSERT_EQ(range->info.last_offset(), base_offset + (i * j) - 1);
+            iobuf output;
+            auto builder = object_builder::create(
+              make_iobuf_ref_output_stream(output), object_builder::options{});
```
It is probably useful to have a test that is close to it's intended usecase: a single object builder, passed to multiple consumers, then verify at the end the incrementally collected state is similar to what the builder recorded.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.h:
```diff
@@ -42,9 +44,8 @@ class reconciler {
       ss::sharded<cluster::partition_manager>*,
-      ss::sharded<cloud_io::remote>*,
       data_plane_api*,
-      std::optional<cloud_storage_clients::bucket_name> = std::nullopt);
+      ss::sharded<l1::file_io>*);
```
for duality with `data_plane_api` consider just taking a raw pointer to the shard local member of this `sharded` class

rockwotj reviewed (commented):
Nice! Just a few smaller comments.

andrwng commented on src/v/cloud_topics/reconciler/range_batch_consumer.cc:
```diff
@@ -14,51 +14,54 @@
-        // Always add term of the first batch in the sequence. We don't really
-        // know if it's the first batch in the term so some filtering will be
-        // needed later.
+    if (_metadata.terms.empty()) {
```
nit: not from this PR, but IMO this is easier to follow as:

```
if (_metadata.terms.empty() || batch.term > _metadata.terms.rbegin()->first) {
    _metadata.terms.insert(batch.term, batch.base_offset());
}
```

The original `!=` here instead of `>` is kinda confusing -- i'm guessing it's to be defensive, but it feels like an edge case that's unlikely to be useful since terms should always increase

andrwng commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -64,35 +56,27 @@ namespace cloud_topics::reconciler {
   ss::sharded<cluster::partition_manager>* pm,
-  ss::sharded<cloud_io::remote>* cloud_io,
   data_plane_api* data_plane,
-  std::optional<cloud_storage_clients::bucket_name> bucket)
+  ss::sharded<l1::file_io>* l1_io)
```
In this, or maybe a subsequent PR, it'd be great to add a fixture test that runs the reconciler calls and does some basic checks about the resulting objects.

For the tiered storage archiver, we have a config property to disable the loop to be able to do this kind of manual testing... which isn't great, but gets the job done..

rockwotj commented:
@wdberkeley Is this ready for another round of reviews?

wdberkeley commented:
@rockwotj Yeah, I revised it. I haven't split it into separate commits like @dotnwat wanted, though.

rockwotj commented on src/v/cloud_topics/reconciler/reconciliation_consumer.cc:
```diff
@@ -0,0 +1,62 @@
+
+namespace cloud_topics::reconciler {
+
+constexpr kafka::offset offset_unset{-1};
```
nit: use `kafka::offset::min()` or `kafka::offset{}` instead.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,93 +157,145 @@ ss::future<> reconciler::reconciliation_loop() {
+    auto output_stream = co_await result.staging_file->output_stream();
+    auto builder = l1::object_builder::create(
+      std::move(output_stream),
+      // TODO: The default is OK?
```
Yeah I think so 🤷

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -64,35 +56,27 @@ namespace cloud_topics::reconciler {
+        [this](ss::lw_shared_ptr<cluster::partition> p) {
+            attach_partition(std::move(p));
+        });
+
```
why is this reordered?

rockwotj reviewed (commented):
LGTM, with some nits. Also you need to rebase as there are conflicts

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,93 +157,145 @@ ss::future<> reconciler::reconciliation_loop() {
-    co_return object;
-}
+    result.object_info = co_await builder->finish();
+    co_await builder->close();
```
This `close` method needs to be called even if any of the above steps throw.

dotnwat reviewed (commented):
is there more than one logically independent change in the PR? if so, it'd be great to split the change into a few separate changes for reviewing.

wdberkeley committed:
cloud_topics/reconciler: Add reconciliation_consumer

Introduces reconciliation_consumer class and tests for consuming record
batches from partitions and writing them to L1 objects. Includes
partition metadata tracking for offsets, timestamps, and term transitions.

wdberkeley committed:
cloud_topics/reconciler: Use local partition manager pointer

Change reconciler constructor to take cluster::partition_manager*
instead of ss::sharded<cluster::partition_manager>*, matching the
pattern used for other local service dependencies, and which will
be used for new reconciler dependencies introduced in a following
commit.

wdberkeley committed:
cloud_topics/reconciler: Remove range_batch_consumer

It's superseded by the reconciliation consumer.

rockwotj commented on src/v/cloud_topics/level_one/common/BUILD:
```diff
@@ -4,13 +4,21 @@ package(default_visibility = ["//src/v/cloud_topics/level_one:__subpackages__"])
     ],
+    visibility = [
+        "//src/v/cloud_topics/level_one:__subpackages__",
+        "//src/v/cloud_topics/reconciler:__subpackages__",
```
change the default visibility at the top of the file isntead

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--85af0405-de85-4371-ae3a-01d13963f418-->
<!--TEST RESULTS-->
<!--7e0d7813-fbe1-4e6c-8ca9-c425913c5b29-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/71937">build#71937</a></summary>


| test_class            | test_method         | test_arguments           | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:----------------------|:--------------------|:-------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_rack_awareness | null                     | integration | https://buildkite.com/redpanda/redpanda/builds/71937#019930c3-90bc-4053-beff-24e9456c59b3 | FLAKY         | 20/21    | upstream reliability is '99.48717948717949'. current run reliability is '95.23809523809523'. drift is 4.24908 and the allowed drift is set to 50. The test should PASS |
| SimpleEndToEndTest    | test_relaxed_acks   | {"write_caching": false} | integration | https://buildkite.com/redpanda/redpanda/builds/71937#019930c3-90c1-4775-afdc-6b73f0750bb8 | FLAKY         | 20/21    | upstream reliability is '98.47715736040608'. current run reliability is '95.23809523809523'. drift is 3.23906 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72012">build#72012</a></summary>


| test_class                   | test_method        | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:-----------------------------|:-------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| PartitionMovementUpgradeTest | test_basic_upgrade | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72012#0199347a-b730-4ec7-a8bb-af65b9b0c2a9 | FAIL          | 0/1      |          |
</details>
<!--/body-->

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,94 +169,178 @@ ss::future<> reconciler::reconciliation_loop() {
+        vlog(
+          lg.error,
+          "Exception building object: {}",
+          object_fut.get_exception());
```
ditto Noah's comment in the other PR. Let's pull this out of the macro invocation.

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,94 +169,178 @@ ss::future<> reconciler::reconciliation_loop() {
+          lg.error,
+          "Exception uploading L1 object {}: {}",
+          object_id,
+          upload_fut.get_exception());
```
nit: same here

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,94 +169,178 @@ ss::future<> reconciler::reconciliation_loop() {
+        auto tidp = ntp_to_topic_id_partition(partition->partition->ntp());
+        if (!tidp.has_value()) {
+            vlog(
+              lg.warn,
```
Let's mark this as an error, there should always be this mapping for cloud topics.

rockwotj reviewed (commented):
This is looking good 😎 

Very close IMO, just a couple nits

wdberkeley committed:
ct/l1: Make level_one/common targets default visible to reconciler

They are all used by the reconciler at this point.

Also removes the reconciler visibility for file_io, as it has
non-default visibility and is no longer used by the reconciler.

wdberkeley committed:
ct/l1: Increase log level and grab exception outside of log macro

rockwotj commented on src/v/cloud_topics/reconciler/reconciler.cc:
```diff
@@ -177,94 +169,171 @@ ss::future<> reconciler::reconciliation_loop() {
+          metadata->last_offset);
+        result.partitions.emplace_back(partition, std::move(metadata.value()));
+
+        auto current_size = co_await staging_file->size();
```
Followup for me to add an api to get this information from the builder. As data could be buffered in the output stream still, this isn't going to give an accurate answer (and we can avoid the syscall)

rockwotj commented on src/v/cloud_topics/reconciler/reconciliation_consumer.h:
```diff
@@ -23,38 +23,31 @@
+struct partition_metadata {
     kafka::offset base_offset;
     kafka::offset last_offset;
     model::timestamp base_timestamp;
```
We don't need the base_timestamp here, and also do we really need this other information? the object builder tracks it and returns the footer when the builder is finished.

wdberkeley replied:
I don't think we'll end up needing it. If it's OK, let's leave it for now as the next change that commits to the metastore fixes this up.

rockwotj replied:
yes sounds good

rockwotj reviewed (approved):
This is a good start for the reconciler!
